File: 54. Spiral Matrix/54. Spiral Matrix.java
import java.util.ArrayList;
import java.util.List;

class Solution {

  public List<Integer> spiralOrder(int[][] matrix) {
    int m = matrix.length, n = matrix[0].length;
    List<Integer> spiral = new ArrayList<>();

    int top = 0, bottom = m - 1;
    int left = 0, right = n - 1;

    while (left <= right && top <= bottom) {
      for (int j = left; j <= right; j++) {
        spiral.add(matrix[top][j]);
      }
      top++;

      for (int i = top; i <= bottom; i++) {
        spiral.add(matrix[i][right]);
      }
      right--;

      if (top <= bottom) {
        for (int j = right; j >= left; j--) {
          spiral.add(matrix[bottom][j]);
        }
        bottom--;
      }

      if (left <= right) {
        for (int i = bottom; i >= top; i--) {
          spiral.add(matrix[i][left]);
        }
        left++;
      }
    }

    return spiral;
  }
}

File: 54. Spiral Matrix/54. Spiral Matrix.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        spiral = []

        top, bottom = 0, m - 1
        left, right = 0, n - 1

        while left <= right and top <= bottom:
            for j in range(left, right + 1):
                spiral.append(matrix[top][j])
            top += 1

            for i in range(top, bottom + 1):
                spiral.append(matrix[i][right])
            right -= 1

            if top <= bottom:
                for j in range(right, left - 1, -1):
                    spiral.append(matrix[bottom][j])
                bottom -= 1

            if left <= right:
                for i in range(bottom, top - 1, -1):
                    spiral.append(matrix[i][left])
                left += 1

        return spiral
